### PR TITLE
index operation example for Motr

### PR DESCRIPTION
**Pull request by huanghua78**
This example shows how to do:
* index create
* index put
* index get
* index delete

/doc/motr-developer-guide.md is updated with the guide.

Signed-off-by: Hua Huang <hua.huang@seagate.com>

### CODE REVIEW

**Comment by huanghua78** [comment]
@johnbent as promised, I added this small example program to show how to write a simple Motr index operation application.

File: doc/motr-developer-guide.md [hunk 1]
```diff
@@ -209,7 +209,67 @@ as the last argument to this program. In this example, we will only use this id
 64-bit of an object identification. This id should be larger than 0x100000ULL, that is 1048576 in decimal.
 
 ## A simple Cortx Motr application (index)
-   * TODO
+Here we will create a simple Cortx Motr application to crate an index, put
```
Comment (andriytk) [inline doc/motr-developer-guide.md:212]: s/crate/create/

File: doc/motr-developer-guide.md [hunk 2]
```diff
@@ -209,7 +209,67 @@ as the last argument to this program. In this example, we will only use this id
...
+some key/value pairs to this index, read key/value pairs back from this index, and then delete it.
```
Comment (andriytk) [inline doc/motr-developer-guide.md:213]: Style: make some use of pronouns:

"... put some key/value pairs to this index, read them back, and then delete the index."

File: motr/examples/example2.c [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+{
+	struct m0_op   *ops[1] = { NULL };
+	struct m0_idx   idx;
+	int             rc = 0;
+
+	m0_idx_init(&idx, &container->co_realm, fid);
+
+	rc = m0_entity_open(&idx.in_entity, &ops[0]) ?:
+	     m0_entity_delete(&idx.in_entity, &ops[0]) ?:
+	     index_launch_wait(&idx.in_entity, ops, rc, NULL);
```
Comment (andriytk) [inline motr/examples/example2.c:105]: op_fini()/free() won't be called here in case of an error.

File: motr/examples/example2.c [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+#include <stdio.h>
+#include <stdlib.h>
+#include <errno.h>
+
+enum {
+	KV_COUNT = 10,
+	KV_LEN   = 32,
+};
+
+static int index_launch_wait(struct m0_entity *e,
```
Comment (andriytk) [inline motr/examples/example2.c:55]: There is nothing index-specific in this function. I'd rather name it like "op_wait_fini()" or something.

File: motr/examples/example2.c [hunk 5]
```diff
@@ -0,0 +1,253 @@
...
+			     struct m0_op    **ops,
+			     int               rc,
```
Comment (andriytk) [inline motr/examples/example2.c:57]: It's quite strange to pass rc as an argument. :)

**Review by andriytk — COMMENTED** [review]
Nod bad! There are a couple of small issues, see above in the comment.

**Review by Saumya-Sunder — APPROVED** [review]

**Comment by Saumya-Sunder** [comment]
Approving for language.

File: doc/motr-developer-guide.md (hunk 1 above)
Comment (huanghua78) [reply doc/motr-developer-guide.md:212]: fixed.

File: doc/motr-developer-guide.md (hunk 2 above)
Comment (huanghua78) [reply doc/motr-developer-guide.md:213]: fixed.

File: motr/examples/example2.c (hunk 3 above)
Comment (huanghua78) [reply motr/examples/example2.c:105]: fixed.

File: motr/examples/example2.c (hunk 4 above)
Comment (huanghua78) [reply motr/examples/example2.c:55]: OK

File: motr/examples/example2.c (hunk 5 above)
Comment (huanghua78) [reply motr/examples/example2.c:57]: Removed.

File: motr/examples/example2.c [hunk 6]
```diff
@@ -52,29 +52,25 @@ enum {
...
+			M0_TIME_NEVER);
+	printf("rc=%d op_rc=%d\n", rc, ops[0]->op_rc);
+	if (sm_rc != NULL) {
+		/* Save retcodes. */
+		*sm_rc = ops[0]->op_rc;
+	}
 	m0_op_fini(ops[0]);
 	m0_op_free(ops[0]);
 	ops[0] = NULL;
 	m0_entity_fini(e);
```
Comment (andriytk) [inline motr/examples/example2.c]: Suggestion: create a separate function for these 4 lines (`op_entity_fini()` or something) and use it here and below at `index_delete()`.

**Review by andriytk — APPROVED** [review]
Looks better now.

File: motr/examples/example2.c (hunk 6 above)
Comment (huanghua78) [reply motr/examples/example2.c]: Done.

**Review by johnbent — APPROVED** [review]

**Review by johnbent — APPROVED** [review]
So awesome!  Great stuff @huanghua78 .  I LOVE how you remembered to link it from the developer guide!

**Pull request merged**